plutus-core/plutus-ir/src/PlutusIR/Transform/EvaluateBuiltins.hs

```diff
@@ -0,0 +1,72 @@
+{-# LANGUAGE LambdaCase          #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE ViewPatterns        #-}
+-- | A pass that tries to evaluate builtin applications in the program.
+--
+-- This functions as a generic constant-folding pass, but which handles
+-- arbitrary builtins.
+module PlutusIR.Transform.EvaluateBuiltins
+    ( evaluateBuiltins
+    ) where
+
+import PlutusCore.Builtin
+import PlutusIR.Contexts
+import PlutusIR.Core
+
+import Control.Lens (transformOf)
+import Data.Functor (void)
+
+evaluateBuiltins
+  :: forall tyname name uni fun a
+  . (ToBuiltinMeaning uni fun
+  , Typeable tyname
+  , Typeable name)
+  => Bool
+  -- ^ Whether to be conservative and try to retain logging behaviour.
+  -> BuiltinVersion fun
+  -> CostingPart uni fun
+  -> Term tyname name uni fun a
+  -> Term tyname name uni fun a
+evaluateBuiltins conservative ver costModel = transformOf termSubterms processTerm
+  where
+    -- Nothing means "leave the original term as it was"
+    eval
+      :: BuiltinRuntime (Term tyname name uni fun ())
+      -> AppContext tyname name uni fun a
+      -> Maybe (Term tyname name uni fun ())
+    eval (BuiltinResult _ getX) AppContextEnd =
+        case getX of
+            MakeKnownSuccess v           -> Just v
+            -- Evaluates successfully, but does logging. If we're being conservative
+            -- then we should leave these in, so we don't remove people's logging!
+            -- Otherwise `trace "hello" x` is a prime candidate for evaluation!
+            MakeKnownSuccessWithLogs _ v -> if conservative then Nothing else Just v
```

Comment:
You could take the result and annotate it with a `trace` call replicating the logging behavior of the builtin application.

Reply:
That's... kind of brilliant. We would then re-evaluate those trace calls every time, which would be kind of annoying.

Also, actually this pass doesn't know about `trace`, it just requires you to be able to evaluate builtins, so that would be a bit less general.

Reply:
> We would then re-evaluate those trace calls every time, which would be kind of annoying.

Sorry, didn't get it.

> Also, actually this pass doesn't know about `trace`, it just requires you to be able to evaluate builtins, so that would be a bit less general.

Right, that's unfortunate.

Reply:
I mean, the optimization will trigger on them every time:
```
trace "hello" x
--> (great, evaluate it) 
x
--> (oops, need to put a trace call in to replicate the logging)
trace "hello" x
--> (great evaluate it)
...
```
etc.

Reply:
Ah, I see. Yeah, if this function knew about `trace`, it could just not touch it. Or you could insert a special annotation and postprocess the term after the optimization is complete. Oh, wait, you can do that right now. If the builtin application emitted a log, you just add it to the annotation of the result and then once you have `trace` available somewhere down the pipeline, you turn it into the actual call to `trace`. Rather stupid design, though.